source/common/common/containers.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "common/common/cleanup.h"
+
+namespace Envoy {
+namespace Common {
+
+/**
+ * Invokes a function for all elements in a container, providing a cleanup function that will be
+ * executed after all the elements have been processed. The Cleanup object is provided to allow each
+ * update callback to delay cleanup until some arbitrary time: the completion callback will be
```

Comment:
The docs on `Cleanup` probably belong inside the method. Probably can just talk about what `done_cb` does and the thread safety. It did require a couple of mins at staring at this to understand what is going on, but I think this is a pretty cool pattern.

Reply:
I updated it a bit, lmk what you think. I think some mention of `Cleanup` is necessary since it's part of the function signature, but I moved most of it into the function body